[JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43603 Benchmarks: temporarily disable profiling-te configuration.
* #43173 [TensorExpr] Start using typecheck in the fuser.
* **#43235 [JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings.**
* #43365 [TensorExpr] Fuser: disallow aten::slice nodes.
* #43171 [TensorExpr] Fuser: only require input shapes to be known (output shapes can be inferred).
* #43170 [TensorExpr] Fuser: move canHandle and some other auxiliary functions into TensorExprFuser class.

This functionality is needed when we want to not lose track of
nodes/values as we merge and unmerge them into other nodes. For
instance, if we have a side data structure with some meta information
about values or nodes, this new functionality would allow to keep that
metadata up to date after merging and unmerging nodes.

Differential Revision: [D23202648](https://our.internmc.facebook.com/intern/diff/D23202648)

Commit:
[JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings.

This functionality is needed when we want to not lose track of
nodes/values as we merge and unmerge them into other nodes. For
instance, if we have a side data structure with some meta information
about values or nodes, this new functionality would allow to keep that
metadata up to date after merging and unmerging nodes.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/utils/subgraph_utils.h:
```diff
@@ -23,11 +23,20 @@ TORCH_API Node* createSingletonSubgraph(Node* n, Symbol subgraphKind);
 // Merge a node into a subgraph node. If `toMerge` is also a subgraph, the
 // subgraphs are merged.
 // `toMerge` is destroyed.
+// An optional argument 'vmap' could be used to retrieve value mappings.
```
Could you specify which direction does the map go in ?

ZolotukhinM replied:
The map is old_values -> new_values. The old values are destroyed as we're merging a node and what this map allows us to do is to locate their existing counterparts (during merging we did create a clone of the original node hence for all old values there exists a counterpart in the `subgraphNode` graph). Does this description help? If yes, I will put it to the comment.

eellison commented on torch/csrc/jit/passes/utils/subgraph_utils.h:
```diff
@@ -23,11 +23,20 @@ TORCH_API Node* createSingletonSubgraph(Node* n, Symbol subgraphKind);
 // `subgraphNode` is destroyed.
+// An optional argument 'vmap' could be used to retrieve value mappings.
 TORCH_API void unmergeSubgraph(Node* subgraphNode);
+TORCH_API void unmergeSubgraph(
```
Same here, could you specify what direction the map goes in?

eellison commented on torch/csrc/jit/passes/utils/subgraph_utils.h:
```diff
@@ -23,11 +23,20 @@ TORCH_API Node* createSingletonSubgraph(Node* n, Symbol subgraphKind);
 // Merge a node into a subgraph node. If `toMerge` is also a subgraph, the
```
Do we not need to update vmap in this case? https://github.com/pytorch/pytorch/pull/43235/files

eellison replied:
Or in this case ? https://github.com/pytorch/pytorch/pull/43235/files#diff-2589d5e7534a9f39ac403b29ff7bb166R23

ZolotukhinM replied:
I'm not sure where these links are pointing to? Is the question about `createSingletonSubgraph` function?

eellison replied:
The only user of merging nodes with blocks afaik is the functional graph stuff which isn’t really being used or maintained, so we could consider removing this behavior

ZolotukhinM replied:
Ah, I see your point. Let me add a TODO so that at least it doesn't get unnoticed - in the long term I agree that it might be worth just removing this functionality if we don't really use it.

eellison commented on torch/csrc/jit/passes/utils/subgraph_utils.cpp:
```diff
@@ -157,6 +200,13 @@ void mergeNodeIntoSubgraph(Node* toMerge, Node* subgraphNode) {
+    vmap[toMerge->input(idx)] = mergedNode->input(idx);
+  }
+
   // If n's outputs were inputs to `group`, remove them since we just merged
```
So, what happens in a case like this: https://github.com/pytorch/pytorch/blob/2c287a20c78b7b7f468f56bfa2a91d080520d709/torch/csrc/jit/passes/utils/subgraph_utils.cpp#L240 ?

If a node is merged into a subgraph, the vmap mapping contains only the mapping from the value to its new value within the subgraph, not its new value in the enclosing graph correct ?

ZolotukhinM replied:
Correct, we don't have a mapping in the enclosing graph. So, we won't notice that a value have been added to inputs/outputs.

I actually thought about some API to detect that a value was registered as a new input/output, but couldn't come up with clean rules for doing so and instead went with the approach presented in this PR.

eellison reviewed (commented):
I'm a little confused by what mapping we're maintaining. Could you clarify and I'll review further ?

I think this might work for your case but is currently confusing as an API... It would be nice to also know how this is being used exactly to think abt if this is the right solution

Krovatkin commented on torch/csrc/jit/passes/utils/subgraph_utils.cpp:
```diff
@@ -11,40 +11,80 @@ bool hasSubgraph(Node* n) {
+
+  // Vmap should contain "original" -> "merged" mapping, thus we basically need
+  // to perform the following transformation:
+  //   vmap[x] = merge_vmap[unmerge_map[x]]
```
nitpick: an extra blank in front of `vmap[x]`

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "[JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings."


This functionality is needed when we want to not lose track of
nodes/values as we merge and unmerge them into other nodes. For
instance, if we have a side data structure with some meta information
about values or nodes, this new functionality would allow to keep that
metadata up to date after merging and unmerging nodes.

Differential Revision: [D23202648](https://our.internmc.facebook.com/intern/diff/D23202648)

[ghstack-poisoned]

Commit:
Update on "[JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings."


This functionality is needed when we want to not lose track of
nodes/values as we merge and unmerge them into other nodes. For
instance, if we have a side data structure with some meta information
about values or nodes, this new functionality would allow to keep that
metadata up to date after merging and unmerging nodes.

Differential Revision: [D23202648](https://our.internmc.facebook.com/intern/diff/D23202648)

[ghstack-poisoned]

Commit:
Update on "[JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings."


This functionality is needed when we want to not lose track of
nodes/values as we merge and unmerge them into other nodes. For
instance, if we have a side data structure with some meta information
about values or nodes, this new functionality would allow to keep that
metadata up to date after merging and unmerging nodes.

Differential Revision: [D23202648](https://our.internmc.facebook.com/intern/diff/D23202648)

[ghstack-poisoned]

Commit:
Update on "[JIT] Subgraph utils: add an optional vmap argument to the API to allow retrieving value mappings."


This functionality is needed when we want to not lose track of
nodes/values as we merge and unmerge them into other nodes. For
instance, if we have a side data structure with some meta information
about values or nodes, this new functionality would allow to keep that
metadata up to date after merging and unmerging nodes.

Differential Revision: [D23202648](https://our.internmc.facebook.com/intern/diff/D23202648)

[ghstack-poisoned]
